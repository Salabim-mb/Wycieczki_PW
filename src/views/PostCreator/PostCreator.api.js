import paths from 'constants/api';

export const getPost = async (token, id) => {
	if (id) {
		// fetch post
		const url = `${paths.BLOG_POST}${id}/`;

		const headers = {};

		const response = await fetch(url, { headers, method: 'GET' });

		if (response.status === 200) {
			return response.json();
		}
		throw new Error('');
	}
	// return empty if id is not provided
	return {};
};

export const uploadPhoto = async (token, id, file, signal, flag, attachment) => {
	const url = `${attachment ? paths.BLOG_DOWNLOADABLE : paths.BLOG_IMAGE}`;

	const formData = new FormData();
	formData.append('reservation', id);
	formData.append(attachment ? 'file' : 'image', file, file.name);

	if (flag) {
		formData.append(flag, true);
	}

	const headers = {
		Authorization: `${token}`,
	};

	const response = await fetch(url, { headers, method: 'POST', signal, body: formData });

	if (response.status === 201) {
		return response.json();
	}
	throw new Error('');
};

export const sendPost = async (token, post, id) => {
	const url = `${paths.BLOG_POST}${id ? `${id}/` : ''}`;

	const headers = {
		Authorization: `${token}`,
		'Content-Type': 'application/json',
	};

	const postNew = { ...post };

	if (post.cover && typeof post.cover !== 'string') {
		const cover = await uploadPhoto('', post.reservation, postNew.cover, null, 'is_cover', false);
		postNew.cover = `${paths.PLAIN}${cover.image_url}`;
	}

	if (post.header && typeof post.header !== 'string') {
		const header = await uploadPhoto(
			'',
			post.reservation,
			postNew.header,
			null,
			'is_header',
			false,
		);
		postNew.header = `${paths.PLAIN}${header.image_url}`;
	}

	const method = id ? 'PATCH' : 'POST';
	const expectedResponseStatus = id ? 200 : 201;

	const response = await fetch(url, { headers, method, body: JSON.stringify(postNew) });

	if (response.status !== expectedResponseStatus) {
		throw new Error('');
	}
};

export const uploadAttachment = async (token, id, file) =>
	uploadPhoto(token, id, file, null, null, true);

export const sendAttachments = async (token, attachmentsObject) => {
	const { attachments, reservation } = attachmentsObject;
	await Promise.all(
		attachments?.map(({ file }) => file && uploadAttachment(token, reservation, file)),
	);
};

export const deleteAttachment = async (token, id, isImage) => {
	const startOfUrl = isImage ? paths.BLOG_IMAGE : paths.BLOG_DOWNLOADABLE;
	const url = `${startOfUrl}${id}/`;

	const headers = {
		Authorization: `${token}`,
	};

	const response = await fetch(url, { headers, method: 'DELETE' });

	if (response.status !== 204) {
		throw new Error('');
	}
	return response;
};

export const deleteAttachments = async (token, files) => {
	const imagesToDelete = files?.images.filter(
		({ image_url: imageUrl }) => !files.content.includes(imageUrl),
	);

	let attachmentsToDelete = [];

	if (files.prevAttachments) {
		attachmentsToDelete = files.prevAttachments.filter(
			e => !files.attachments.some(item => item.id === e.id),
		);
	}

	await Promise.all(imagesToDelete.map(({ id }) => deleteAttachment(token, id, true)));

	await Promise.all(attachmentsToDelete.map(({ id }) => deleteAttachment(token, id, false)));
};

export const reserveSpace = async token => {
	const url = `${paths.BLOG_RESERVATION}`;

	const headers = {
		Authorization: `${token}`,
	};

	const response = await fetch(url, { headers, method: 'POST' });

	if (response.status === 201) {
		return response.json();
	}
	throw new Error('');
};

export const getReservation = async id => {
	if (id) {
		const url = `${paths.BLOG_RESERVATION}${id}/`;

		const headers = {};

		const response = await fetch(url, { headers, method: 'GET' });

		if (response.status === 200) {
			return response.json();
		}
		throw new Error('');
	}
	// return empty if id is not provided
	return {};
};

export const getTopics = async () => {
	const url = `${paths.BLOG_TOPICS}`;

	const headers = {};

	const response = await fetch(url, { headers, method: 'GET' });

	if (response.status !== 200) {
		throw new Error('błąd');
	}

	return response.json();
};
