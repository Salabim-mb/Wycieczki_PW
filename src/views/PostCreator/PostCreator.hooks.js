import { useMutation, useQuery } from 'react-query';
import {
	sendPost,
	sendAttachments,
	getTopics,
	deleteAttachments,
	deleteAttachment,
} from 'views/PostCreator/PostCreator.api';

export const usePostMutation = (token, prevId) =>
	useMutation(postData => sendPost(token, postData, prevId));
export const useAttachmentsMutation = token =>
	useMutation(attachmentList => sendAttachments(token, attachmentList));
export const useTopicsQuery = () => useQuery('topics', getTopics);
export const useFileDeletion = token => useMutation(data => deleteAttachments(token, data));
export const useImageDeletion = token => useMutation(data => deleteAttachment(token, data, true));
