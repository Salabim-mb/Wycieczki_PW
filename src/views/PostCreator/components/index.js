export { default as PostForm } from './PostForm';
export { default as ImageAdapter } from './ImageAdapter';
export { default as TopicSelector } from './TopicSelector';
export { default as AttachmentsList } from './AttachmentsList';
export { default as PostTilePreview } from './PostTilePreview';
export { default as HeaderPreview } from './HeaderPreview';
export { default as HeaderInput } from './HeaderInput';
export { default as PostEditor } from './PostEditor';
