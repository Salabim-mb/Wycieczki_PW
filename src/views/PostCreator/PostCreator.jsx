import React from 'react';
import { useParams } from 'react-router-dom';
import { AlertInfo } from 'components';
import { Container, Card, CardHeader, CardContent } from '@material-ui/core';

import { PostForm } from './components';
import { usePostQuery } from './PostCreator.hooks';

const PostCreator = () => {
	const { id } = useParams();

	const { isLoading, isError, error, data } = usePostQuery(id);

	return (
		<Container>
			<Card>
				<CardHeader title="StwÃ³rz wpis" />
				<CardContent>
					<PostForm post={data && Object.keys(data).length === 0 ? null : data} id={id} />
					<AlertInfo isLoading={isLoading} isError={isError}>
						{error?.message}
					</AlertInfo>
				</CardContent>
			</Card>
		</Container>
	);
};

export default PostCreator;
