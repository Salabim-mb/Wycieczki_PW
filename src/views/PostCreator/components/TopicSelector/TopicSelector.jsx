import React from 'react';
import { Select, MenuItem } from '@material-ui/core';
import PropTypes from 'prop-types';
import { useQuery } from 'react-query';
import { AlertInfo } from 'components';
import { getTopics } from 'views/PostCreator/PostCreator.api';

const TopicSelector = ({ value, handleChange }) => {
	const { isLoading, isError, error, data, isSuccess } = useQuery('topics', getTopics, {
		enabled: false,
	});
	return (
		<>
			<Select labelId="topicSelector" id="select" value={value} onChange={handleChange}>
				<MenuItem value={-1} disabled>
					Wybierz temat
				</MenuItem>
				{isSuccess &&
					data.map(({ id, title }) => (
						<MenuItem key={id} value={id}>
							{title}
						</MenuItem>
					))}
			</Select>
			<AlertInfo isLoading={isLoading} isError={isError}>
				{error?.message}
			</AlertInfo>
		</>
	);
};

TopicSelector.propTypes = {
	value: PropTypes.number.isRequired,
	handleChange: PropTypes.func.isRequired,
};

export default TopicSelector;
