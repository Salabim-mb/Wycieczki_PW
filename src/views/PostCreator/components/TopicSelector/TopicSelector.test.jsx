import React from 'react';
import { render } from '@testing-library/react';
import TopicSelector from './TopicSelector';
import { useTopicsQuery } from '../../PostCreator.hooks';

jest.mock('views/PostCreator/PostCreator.hooks', () => ({
	useTopicsQuery: jest.fn(),
}));

describe('TopicSelector', () => {
	let props;

	beforeEach(() => {
		jest.clearAllMocks();

		props = { value: -1, handleChange: jest.fn() };
		useTopicsQuery.mockImplementation(() => ({
			isSuccess: true,
			data: [
				{
					id: 1,
					title: 'Kolorowanki',
				},
				{
					id: 2,
					title: 'Kolorowanki2',
				},
			],
			isError: false,
			isLoading: false,
		}));
	});

	it('should match snapshot', async () => {
		const { container } = render(<TopicSelector {...props} />);

		expect(container).toMatchSnapshot();
	});
});
