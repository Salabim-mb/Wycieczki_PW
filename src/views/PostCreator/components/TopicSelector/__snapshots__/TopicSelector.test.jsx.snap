// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopicSelector should match snapshot 1`] = `
<div>
  <div
    class="MuiInputBase-root MuiInput-root MuiInput-underline"
  >
    <div
      aria-haspopup="listbox"
      aria-labelledby="topicSelector select"
      class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
      id="select"
      role="button"
      tabindex="0"
    >
      Wybierz temat
    </div>
    <input
      aria-hidden="true"
      class="MuiSelect-nativeInput"
      tabindex="-1"
      value="-1"
    />
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSelect-icon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M7 10l5 5 5-5z"
      />
    </svg>
  </div>
</div>
`;
